### PR TITLE
drivers: versal: true random number generator

### PR DESCRIPTION
**Pull request by ldts**
Configure the TRNG driver to operate in Hybrid mode with derivative
function.

This driver was ported from its original FSBL implementation (1)

(1) https://github.com/Xilinx/embeddedsw

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

reference: https://github.com/OP-TEE/optee_os/pull/5426

### CODE REVIEW

**Review by etienne-lms — COMMENTED** [review]
some comments. LGTM otherwise.

**Comment by ldts** [comment]
@etienne-lms thanks a lot for your comments!. will update the PR in the morning (sorry for the late responses but you saw all the other stuff pending at the same time)

File: core/drivers/versal_trng.c [hunk 1]
```diff
@@ -0,0 +1,1151 @@
...
+ * approved RBG or an entropy source provides full entropy output when entropy
+ * input is requested by the DRBG mechanism. Otherwise, the derivation function
+ * shall be used". Sufficient large entropy data from user is fed to DF to
+ * generate the seed which will be loaded into the external seed registers.
+ * From here, it is similar to regular DRNG mode.
+ *
+ * PTRNG mode with DF: This mode is similar to PTRNG mode, however, the entropy
+ * data from the core output registers are accumulated and fed to the DF
+ * (instead of directly consuming it). The output of the DF would be final
+ * random data. In this mode, the output of DF is not seed but the random data.
```
[the remark this answers is not in the document]
Comment (ldts) [reply core/drivers/versal_trng.c:48]: I think it is right: usually the DF is used to just generate a seed. The comment is saying that when used in this mode, the DF output is not a seed but the actual random data.

**Comment by ldts** [comment]
@etienne-lms updated and rebased

**Comment by ldts** [comment]
@etienne-lms maybe it can be merged?

File: core/drivers/versal_trng.c [hunk 2]
```diff
@@ -0,0 +1,1151 @@
...
+	return TEE_SUCCESS;
+}
+
+void plat_rng_init(void)
+{
+}
+
+static TEE_Result trng_hrng_mode_init(void)
+{
+	uint8_t pers_str[TRNG_PERS_STR_LEN] = {
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/drivers/versal_trng.c:1093]: @ldts this comment wasn't addressed

[the remark this answers is not in the document]
Comment (ldts) [reply core/drivers/versal_trng.c]: @jforissier thanks, didnt notice your comment earlier.... just pushed it.

File: core/drivers/versal_trng.c [hunk 3]
```diff
@@ -0,0 +1,1147 @@
...
+#define TRNG_STATUS_DONE_MASK		BIT(0)
+#define TRNG_CTRL			0x08
+#define TRNG_CTRL_EUMODE_MASK		BIT(8)
+#define TRNG_CTRL_PRNGMODE_MASK		BIT(7)
+#define TRNG_CTRL_PRNGSTART_MASK	BIT(5)
+#define TRNG_CTRL_PRNGXS_MASK		BIT(3)
+#define TRNG_CTRL_TRSSEN_MASK		BIT(2)
+#define TRNG_CTRL_PRNGSRST_MASK		BIT(0)
+#define TRNG_EXT_SEED_0			0x40
+/* Below registers are not directly referenced in driver but are accessed
```
Comment (etienne-lms) [inline core/drivers/versal_trng.c:85]:
```
/*
 * ...
 *
```
Ditto at line 100.

File: core/drivers/versal_trng.c [hunk 4]
```diff
@@ -0,0 +1,1147 @@
...
+		xfer_len = DF_PAD_DATA_LEN + TRNG_PERS_STR_LEN;
+		src = (uintptr_t)trng->dfin.pstr;
+		offset = MAX_PRE_DF_LEN - trng->len;
+	}
+
+	/* Move back into the dfin structure */
+	dst = src - offset;
+	reminder = (uintptr_t)&trng->dfin + sizeof(trng->dfin) - offset;
+	if (offset) {
+		memcpy((void *)dst, (void *)src, xfer_len);
```
Comment (etienne-lms) [inline core/drivers/versal_trng.c:456]: `memove()` seems a better option.

File: core/drivers/versal_trng.c [hunk 5]
```diff
@@ -0,0 +1,1147 @@
...
+
+	if (trng_reseed(&versal_trng, NULL, usr_cfg.dfmul)) {
+		EMSG("Reseed Failed");
+		goto error;
+	}
+
+	return TEE_SUCCESS;
+error:
+	panic();
+	return TEE_ERROR_GENERIC;
```
Comment (etienne-lms) [inline core/drivers/versal_trng.c:1144]: Remove the return instruction.

Maybe replace `goto error` with `panic` straight? Whatever, do as you prefer.

**Review by etienne-lms — COMMENTED** [review]
Comment on commit message: fix link using a `Link:` tag.
```
This driver was ported from its original FSBL implementation [1].

[1] https://github.com/Xilinx/embeddedsw
```

**Comment by etienne-lms** [comment]
Hi @ldts, sorry i was a bit busy and left P-R reviews for a while and only report now these 2 coding style issues (inline comments) and must importantly a `memcpy()` that should be replaced with a `memmove()`.

File: core/drivers/versal_trng.c (hunk 4 above)
Comment (ldts) [reply core/drivers/versal_trng.c:456]: um, why, how do you know src is not being used somewhere else. I dont feel comfortable with this change. why do you feel it is required?

Comment (etienne-lms) [reply core/drivers/versal_trng.c:456]: src and dst are in the same address range. As stated in the inline comment, this move data inside a single buffer, hence this should use `memmove()` IIUC.

Comment (etienne-lms) [reply core/drivers/versal_trng.c:456]: yes, i was thinking of overlaps. From memcpy() man page:
> DESCRIPTION
>      The  memcpy()  function  copies n bytes from memory area src to memory area dest.
>      **The memory areas must not overlap.  Use memmove(3) if the memory areas do overlap.**

Maybe `Xil_SMemCpy()` in the referred code ensures destination address is lower than source when buffer do overlap. Using `memcpy()` you do not have this ensurance.

Comment (ldts) [reply core/drivers/versal_trng.c:456]: um, I think you are right.
maybe we should just do likewise?
```
int Xil_SMemCpy(void *Dest, const u32 DestSize,
	const void *Src, const u32 SrcSize, const u32 CopyLen)
{
	int Status = XST_FAILURE;
	const u8 *Src8 = (const u8 *) Src;
	const u8 *Dst8 = (u8 *) Dest;
	void * volatile DestTemp = Dest;
	const void * volatile SrcTemp = Src;

	if ((Dest == NULL) || (Src == NULL)) {
		Status =  XST_INVALID_PARAM;
	}
	else if ((CopyLen == 0U) || (DestSize < CopyLen) || (SrcSize < CopyLen)) {
		Status =  XST_INVALID_PARAM;
	}
	/* Return error for overlap string */
	else if ((Src8 < Dst8) && (&Src8[CopyLen - 1U] >= Dst8)) {
		Status =  XST_INVALID_PARAM;
	}
	else if ((Dst8 < Src8) && (&Dst8[CopyLen - 1U] >= Src8)) {
		Status =  XST_INVALID_PARAM;
	}
	else {
		(void)memcpy(DestTemp, SrcTemp, CopyLen);
		Status = XST_SUCCESS;
	}

	return Status;
}
```

Comment (ldts) [reply core/drivers/versal_trng.c:456]: I think when I rewrote this code early this year I already made all the proper corrections thought so we are probably are going in circles...but it is a good thing to review.

Comment (ldts) [reply core/drivers/versal_trng.c:462]: just pushed a safeguard (in any case the algorithm expected the remainder to be set to 0 so truncating seems to be fine.

**Comment by jforissier** [comment]
For "MAINTAINERS: add Versal ACAP TRNG maintainer":
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by ldts** [comment]
> Hi @ldts, sorry i was a bit busy and left P-R reviews for a while and only report now these 2 coding style issues (inline comments) and must importantly a `memcpy()` that should be replaced with a `memmove()`.

hi thanks! @etienne-lms I followed up where you made the original comments. safeguarding against a potential overlap seems a better choice as we need to remain in the structure.

**Comment by ldts** [comment]
done then?

**Comment by ldts** [comment]
and rebased

File: core/drivers/versal_trng.c [hunk 6]
```diff
@@ -0,0 +1,1150 @@
...
+	trng->status = TRNG_ERROR;
+	return TEE_ERROR_GENERIC;
+}
+
+static struct versal_trng versal_trng = {
+	.cfg.base = TRNG_BASE,
+	.cfg.len = TRNG_SIZE,
+};
+
+TEE_Result hw_get_random_bytes(void *buf, size_t len)
```
Comment (zjuzzy) [inline core/drivers/versal_trng.c:1070]: Does it need a header file to declare this interface? so that others can use this interface.

Comment (ldts) [reply core/drivers/versal_trng.c:1070]: no it is not required. the interface is already in core/include/rng_support.h (it is not driver specific)

**Comment by ldts** [comment]
so all good? @jforissier ?

**Comment by jforissier** [comment]
@ldts LGTM you may add:
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

But I'd like @etienne-lms's ack or review tag since he did most of the review.

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` with proposal on remaining comment applied.

File: core/drivers/versal_trng.c [hunk 7]
```diff
@@ -0,0 +1,1150 @@
...
+		offset = MAX_PRE_DF_LEN - trng->len;
+	}
+
+	/* Move back into the dfin structure */
+	dst = src - offset;
+	reminder = (uintptr_t)&trng->dfin + sizeof(trng->dfin) - offset;
+	if (offset) {
+		/* safeguard memcpy operation */
+		if (xfer_len > offset)
+			xfer_len = offset;
```
[the remark this answers is not in the document]
Comment (ldts) [reply core/drivers/versal_trng.c:460]: No that is ok, I'll panic and configure the default multiplier as per the reference code. thanks!

**Comment by ldts** [comment]
@jforissier @etienne-lms thanks for the review. updated!

**Pull request merged**